components/brave_wallet/browser/solana_tx_manager.cc

```diff
@@ -85,19 +87,189 @@ void SolanaTxManager::AddUnapprovedTransaction(
+    estimation->compute_units = compute_units;
+    estimation->fee_per_compute_unit = kMininumFeePerComputeUnits;
+    std::move(callback).Run(std::move(meta), std::move(estimation), error,
+                            error_message);
+    return;
+  }
+
+  uint64_t median = 0;
+  if (!recent_fees.empty()) {
+    std::sort(recent_fees.begin(), recent_fees.end(),
+              [](const std::pair<uint64_t, uint64_t>& a,
+                 const std::pair<uint64_t, uint64_t>& b) {
+                return a.second < b.second;
+              });
+
+    size_t size = recent_fees.size();
+    if (size % 2 == 0) {
+      median =
+          (recent_fees[size / 2 - 1].second + recent_fees[size / 2].second) / 2;
+    } else {
+      median = recent_fees[size / 2].second;
+    }
+  }
+
+  mojom::SolanaFeeEstimationPtr estimation = mojom::SolanaFeeEstimation::New();
+  estimation->base_fee = base_fee;
+  // The simulation was performed without the instructions that set a compute
+  // budget and priority fee, so we must add those as well.
+  estimation->compute_units = compute_units + kAddPriorityFeeComputeUnits;
+
+  if (median == 0) {
+    estimation->fee_per_compute_unit = kMininumFeePerComputeUnits;
+  } else {
+    estimation->fee_per_compute_unit = median;
+  }
+
+  std::move(callback).Run(std::move(meta), std::move(estimation),
+                          mojom::SolanaProviderError::kSuccess, "");
+}
+
+void SolanaTxManager::OnGetEstimatedTxBaseFee(
+    const std::string& chain_id,
+    GetSolanaTxFeeEstimationAndMetaCallback callback,
+    std::unique_ptr<SolanaTxMeta> meta,
+    uint64_t base_fee,
+    mojom::SolanaProviderError error,
+    const std::string& error_message) {
+  if (error != mojom::SolanaProviderError::kSuccess) {
+    std::move(callback).Run({}, {}, error, error_message);
+    return;
+  }
+
+  const auto unsigned_tx = meta->tx()->GetUnsignedTransaction();
+  auto internal_callback =
+      base::BindOnce(&SolanaTxManager::OnSimulateSolanaTransaction,
+                     weak_ptr_factory_.GetWeakPtr(), chain_id, std::move(meta),
+                     base_fee, std::move(callback));
+  json_rpc_service_->SimulateSolanaTransaction(chain_id, unsigned_tx,
+                                               std::move(internal_callback));
+}
+
+void SolanaTxManager::FinishGetSolanaTxFeeEstimation(
+    GetSolanaTxFeeEstimationCallback callback,
+    std::unique_ptr<SolanaTxMeta> meta,
+    mojom::SolanaFeeEstimationPtr estimation,
+    mojom::SolanaProviderError error,
+    const std::string& error_message) {
+  std::move(callback).Run(std::move(estimation), error, error_message);
+}
+
+void SolanaTxManager::ContinueAddUnapprovedTransaction(
+    AddUnapprovedTransactionCallback callback,
+    std::unique_ptr<SolanaTxMeta> meta,
+    mojom::SolanaFeeEstimationPtr estimation,
+    mojom::SolanaProviderError error,
+    const std::string& error_message) {
+  // If the base fee is zero (fetching failed), we add the transaction without
```

Comment:
I'm a bit confused regarding the `base fee is zero` case, which fetch is this referring to, base fee fetch failed or priority fee fetch failed?
From the comment below it seems we're saying the base fee fetching failed here, but it also seems if fetching base fee failed, we would already run the callback in `OnGetEstimatedTxBaseFee`?
And should this be || instead of && if this is referring to base fee fetch failed?

Reply:
>I'm a bit confused regarding the base fee is zero case, which fetch is this referring to, base fee fetch failed or priority fee fetch failed?

I mean to refer to the case when the base fee fetch failed

> From the comment below it seems we're saying the base fee fetching failed here, but it also seems if fetching base fee failed, we would already run the callback in OnGetEstimatedTxBaseFee?

I believe the callback called in OnGetEstimatedTxBaseFee is an inner callback of type GetSolanaTxFeeEstimationAndMetaCallback (ContinueAddUnapprovedTransaction). So it calls ContinueAddUnapprovedTransaction with an empty fee estimation + the error, and estimation.

> And should this be || instead of && if this is referring to base fee fetch failed?

In the case that base fee fetching succeeded, but we get an error fetching the priority fee, then `error != mojom::SolanaProviderError::kSuccess` and base_fee to be non zero. In that case, we still want to add the gas estimation (for the sake of preserving the base_fee for the client).

Reply:
If I'm reading correctly, if there was an error fetching tx base fee,`std::move(callback).Run({}, {}, error, error_message);` in `OnGetEstimatedTxBaseFee` will be run. Then we will have null estimation instead of a 0 base fee?

Reply:
In any cases, accessing estimation directly in this function without check if it's null first is unsafe.

Reply:
I added null checking, but also altered the logic in OnGetEstimatedTxBaseFee so it does pass a non null SolanaGasEstimationPtr, so that the above logic can be used.